File: grooves-java/src/test/resources/domains/Transaction.java
package domains;

import static rx.Observable.empty;
import static rx.Observable.just;
import static rx.RxReactiveStreams.toPublisher;

import com.github.rahulsom.grooves.api.events.BaseEvent;
import com.github.rahulsom.grooves.api.events.RevertEvent;
import java.util.Date;
import org.reactivestreams.Publisher;

public abstract class Transaction implements BaseEvent<Account, Long, Transaction> {
    private Account aggregate;
    private RevertEvent<Account, Long, Transaction> revertedBy;
    private Long id;
    private long position;
    private Date timestamp;

    @Override
    public void setPosition(long position) {
        this.position = position;
    }

    @Override
    public long getPosition() {
        return position;
    }

    @Override
    public Long getId() {
        return id;
    }

    @Override
    public void setRevertedBy(RevertEvent<Account, Long, Transaction> revertedBy) {
        this.revertedBy = revertedBy;
    }

    @Override
    public RevertEvent<Account, Long, Transaction> getRevertedBy() {
        return revertedBy;
    }

    @Override
    public void setTimestamp(Date timestamp) {
        this.timestamp = timestamp;
    }

    @Override
    public Date getTimestamp() {
        return timestamp;
    }

    @Override
    public void setAggregate(Account aggregate) {
        this.aggregate = aggregate;
    }

    @Override
    public Account getAggregate() {
        return aggregate;
    }

    public Publisher<Account> getAggregateObservable() {
        return toPublisher(aggregate != null ? just(aggregate) : empty());
    }
}

File: grooves-example-pushstyle/src/main/kotlin/grooves/example/push/Transaction.kt
package grooves.example.push

import com.github.rahulsom.grooves.api.events.BaseEvent
import com.github.rahulsom.grooves.api.events.RevertEvent
import io.reactivex.Flowable
import org.reactivestreams.Publisher
import java.util.Date

sealed class Transaction(
    override val id: String,
    override var aggregate: Account?,
    override var timestamp: Date,
    override var position: Long
) :
    BaseEvent<Account, String, Transaction> {

    override fun getAggregateObservable(): Publisher<Account> =
        Flowable.fromIterable(listOfNotNull(aggregate))

    override var revertedBy: RevertEvent<Account, String, Transaction>?
        get() = revertedBy
        set(value) {
            revertedBy = value
        }

    data class Deposit(
        override val id: String,
        override var aggregate: Account?,
        override var timestamp: Date,
        override var position: Long,
        val atmId: String,
        val amount: Long
    ) :
        Transaction(id, aggregate, timestamp, position)

    data class Withdraw(
        override val id: String,
        override var aggregate: Account?,
        override var timestamp: Date,
        override var position: Long,
        val atmId: String,
        val amount: Long
    ) :
        Transaction(id, aggregate, timestamp, position)
}